BUG: Matched SearchedSorted Signature with NumPy's

gfyoung opened the pull request:
Title is self-explanatory.  Closes #12238.

Commit:
BUG: Matched searchsorted signature with numpy's

Closes gh-12238.

gfyoung commented:
Where should I put tests for this compatibility issue?  It would be great if I could just write one test for all of the classes that have a ``searchsorted`` method.

jreback commented:
test_base will exercise this

gfyoung commented:
Added test, and the test passes on Travis.  Should be good to merge if there is nothing else.

gfyoung commented:
I need to get a new mouse.  In any case, I did let the tests run with my new test, and Travis gave the green light.

jreback commented on pandas/tests/test_base.py:
```diff
@@ -1004,6 +1004,16 @@ def f():
+    # See gh-12238
+    def testSearchSorted(self):
+        mixin = IndexOpsMixin()
+        mixin.values = np.array([1, 2, 3])
```
no, see how we do for example ``.value_counts`` where an object is constructed and looped thru. pls follow the same pattern

gfyoung commented:
Tests are finally passing!  Should be good to merge if there is nothing else.

jreback commented on doc/source/whatsnew/v0.18.0.txt:
````diff
@@ -1117,3 +1117,4 @@ Bug Fixes
 - Bug in ``DataFrame.apply`` in which reduction was not being prevented for cases in which ``dtype`` was not a numpy dtype (:issue:`12244`)
 - Bug when initializing categorical series with a scalar value. (:issue:`12336`)
 - Bug when specifying a UTC ``DatetimeIndex`` by setting ``utc=True`` in ``.to_datetime`` (:issue:11934)
+- Bug in ``searchsorted`` argument signatures for ``IndexOpsMixin``, ``DatetimeIndex``, ``PeriodIndex``, and ```TimedeltaIndex`` that broke compatibility with ``searchsorted`` in ``numpy`` (:issue:`12238`)
````
you don't need to say all this, just say ``searchsorted`` now accepts ``sorter`` arg for compat with numpy.

gfyoung replied:
Done.

jreback commented on pandas/core/base.py:
```diff
@@ -990,7 +990,7 @@ def factorize(self, sort=False, na_sentinel=-1):
         return factorize(self, sort=sort, na_sentinel=na_sentinel)
 
-    def searchsorted(self, key, side='left'):
+    def searchsorted(self, key, side='left', sorter=None):
```
actually I would do this like we do ``_validate_kwargs``, e.g. add ``**kwargs`` to the signature and accept ``sorter``, but raise on other args. The reason is we don't actually use this arg (and therefore are not documenting it). you *may* want/need to move ``_validate_kwargs`` to ``core.common`` (or similar)

gfyoung replied:
I don't believe so.  If you look at the `numpy` call to `searchsorted`, it unfortunately passes them in as if they were positional arguments.

jreback replied:
really? that is odd, ok then pls add to the doc-strings where you added it with a comment that this is for compatiblity only.

Further that would be a bug in numpy. It shouldn't be passing positional arguments to sub-class likes.

jreback commented on pandas/tests/test_base.py:
```diff
@@ -972,6 +972,18 @@ def test_memory_usage(self):
             self.assertTrue(abs(diff) < 100)
 
+    def test_search_sorted(self):
+        # See gh-12238
```
why are you doing: ``mixin.values = o``? no reason for this

jreback commented on pandas/tseries/period.py:
```diff
@@ -387,7 +387,7 @@ def astype(self, dtype):
         raise ValueError('Cannot cast PeriodIndex to dtype %s' % dtype)
 
-    def searchsorted(self, key, side='left'):
+    def searchsorted(self, key, side='left', sorter=None):
```
same here

jreback commented on pandas/tseries/tdi.py:
```diff
@@ -784,7 +784,7 @@ def _partial_td_slice(self, key, freq, use_lhs=True, use_rhs=True):
         # return (lhs_mask & rhs_mask).nonzero()[0]
 
-    def searchsorted(self, key, side='left'):
+    def searchsorted(self, key, side='left', sorter=None):
```
same here

jreback commented:
you can put in 0.18.0

gfyoung commented:
You accidentally cancelled all of my builds for this PR.

jreback commented:
oh sorry, push again :)

gfyoung commented:
FYI: I believe you can restart a build, but no worries here since I just rebased onto `master` as my new "commit".

jreback commented:
right, but just as easy to:

``git commit --all --amend -C HEAD`` to generate a new commit as well.

gfyoung commented:
Ah, good to know!

jreback commented on pandas/core/base.py:
```diff
@@ -990,11 +990,17 @@ def factorize(self, sort=False, na_sentinel=-1):
 
-    def searchsorted(self, key, side='left'):
-        """ np.ndarray searchsorted compat """
+    def searchsorted(self, key, side='left', sorter=None):
```
just realized that ``Series.searchsorted`` has a doc-string (and includes the ``sorter`` arg, so:

can you move the ``Series.searchsorted`` doc-string to ``pandas/core/base.py`` (and update it with the warning that you wrote below).

Then since everything else inherits this, I think ou can drop the docs strings (and they will be inherited).

gfyoung replied:
1) Does `Series` not do something explicit with `sorter` in its implementation?
2) Unfortunately, I don't believe the doc strings are inherited at the moment (i.e. on `master`) because if you type `help(DatetimeIndex.searchsorted)` and `help(IndexOpsMixin.searchsorted)`, the former returns an string, while the other returns `np.ndarray searchsorted compat`.

jreback replied:
doc strings ARE inherited if the subs don't have doc strings (but they do now) that's why u don't see it 

the sorter is passed onto np.searchsorter I think

ok so in that case - you can prob futz with it a bit and just pass theu sorter

gfyoung replied:
Actually, `DatetimeIndex.searchsorted` has no doc string at the moment.  See <a href="https://github.com/pydata/pandas/pull/12413/files?diff=unified#diff-23ecb29e7ceba52109a365e447400d2eL1630">here</a>

jreback replied:
so u prob have to use Apoenders and s shared doc string then

we don't want to duplicate things like they are now

jreback replied:
just an FYI

we have a complete opposite policy as compared to numpy 

we go to great lengths to avoid duplication of code (doc strings whatever)

I find that just leads to s great many inconsistencies (the duplication)

gfyoung replied:
Could you clarify your last comment?  I don't quite understand "Apoenders"

gfyoung replied:
Could you clarify your last comment?  I don't quite understand "Apoenders"

jreback replied:
looks for the Appender decorator and also Substitute decorator
used to propagate doc strings 
good examples in core/generic then usage in core/Frame

eg reindex

gfyoung replied:
Got it.  Done.

jreback commented on pandas/tseries/tdi.py:
```diff
@@ -784,7 +784,15 @@ def _partial_td_slice(self, key, freq, use_lhs=True, use_rhs=True):
         # # try to find a the dates
         # return (lhs_mask & rhs_mask).nonzero()[0]
 
-    def searchsorted(self, key, side='left'):
```
also you can fix the signatures here to match (e.g. ``key`` -> ``v``)

gfyoung replied:
Done.

shoyer commented:
We should raise `NotImplementedError` if sorter is not None. It's not a good idea to insert arguments without using them.

gfyoung commented:
@shoyer : I'm a little wary of doing that because it breaks compatibility with `numpy`.

kawochen commented:
I agree with @shoyer.  I don't think it would break compatibility since it's just validation.

shoyer commented:
NumPy needs to be able to pass the `sorter` argument on to pandas so the `np.searchsorted` wrapper works. But the default sorter is `None`, which is the same default we'll use here for the pandas method. Given that we don't support other values for `sorter`, we should raise an error if that happens.

gfyoung commented:
Fair enough.  `NotImplementedError` added.

gfyoung commented:
Tests are passing once more.  Should be good to merge if there is nothing else.

jreback commented on pandas/core/base.py:
```diff
@@ -15,6 +16,24 @@
+    for a given class instance 'obj' should accept a non-None
+    'sorter' argument, if one is provided. Raises a
+    NotImplementedError if that is not the case.
+    """
```
this should be part of the ``searchsorted`` method, then just call ``super(...)`` in the subs. Again this is a difference from numpy, where we tend to have as much possible code in the same place, and avoid a litterring of functions. You can always just raise on ``sorter is not None``. doesn't matter what it is. We don't test this. If its an issue in the future it can then be implementd.

gfyoung replied:
I'm not sure why I would want to call `super`.  Doesn't that mean I end doing search twice?

jreback replied:
of course not. you need to put the actual impl here as well. The point of these cleanups is to have less, more concise, more easily understandable code. Not more functions which are scattered.

gfyoung replied:
I don't think I quite follow what you are saying.  What do you mean by "you need to put the actual impl here as well"?  I'm sure what you are saying makes sense in conception, but it is not coming across clearly to me at the moment on how to proceed with regards to the organization.

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +1009,94 @@ def factorize(self, sort=False, na_sentinel=-1):
+
+        See Also
+        --------
+        TimedeltaIndex.searchsorted
```
just list ``numpy.searchsorted`` the rest is superfluous

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +1009,94 @@ def factorize(self, sort=False, na_sentinel=-1):
+            If 'left', the index of the first suitable location found is given.
+            If 'right', return the last such index.  If there is no suitable
+            index, return either 0 or N (where N is the length of `a`).
+        sorter : 1-D array_like, optional
```
I either wouldn't even put this here at all, or put a comment that is for compatiblity with numpy; we are not advertising this at all. To be honest I really just like the strategy of not even having this in the signature, but allowing it to be passed (but must be ``None``). that way it doesn't need to be documented and is not confusing to the user.

We don't want to have keywords that are there just because numpy has them.

jreback replied:
IOW, use ``_validate_kwargs``

gfyoung replied:
But the `Series` implementation uses the `sorter` argument.  This is what makes it slightly tricky here.  As much as I would like to "share" code, they all don't behave in the same way.  Also, `numpy` passes these arguments as position arguments currently (I'm trying to change that in my recent PR against the `numpy` library) so `_validate_kwargs` can't be used ATM.

jreback replied:
how does it *use* the ``sorter`` arg? are there tests where it is actually used? is it useful?

gfyoung replied:
The `sorter` arg specifies the `order` in which the "sorted array" is to be searched.  There aren't any tests using it for `Series` but there are for `Categorical` in `tests/test_categorical.py`.

jorisvandenbossche replied:
Can't we just support `sorter` in Index as well? (it's just passing the argument to the numpy method)? That would simplify things

gfyoung replied:
I agree.  However, @shoyer said though that externalizing an argument such as `sorter` and then not using it was not really a good idea, hence why the `NotImplementedError` was used in `Index`.  Personally, I am not entirely sure which is the "proper" way to go, as either make sense.  Ideally, we would want to swallow it into `kwargs` arg so there is no externalization but that is not possible given that `numpy` calls the function internally with positional arguments.

jreback replied:
you can do exactly the same thing with ``*args`` (which in this case must be either be 0 len or must be len of 1 with a value of ``None``).

I really don't want this argument exposed simply because numpy is doing something dumb and passing it on.

gfyoung replied:
@jreback : What is your opinion on this issue?  There seems to be a divide amongst people here regarding whether or not we should fully "accept" `sorter` as a keyword argument to `searchsorted`.

gfyoung replied:
@jreback : Oh, that's an interesting idea.  Do we need to specify that a len-1 `*args` be `None`?  I wonder if we could do the same thing as we do with `_validate_kwargs`, which is check to see if  `numpy` compatible arguments have been passed in and ignore them otherwise.

gfyoung replied:
Which in this case would mean: is `len(args) <= 1`?  Perhaps we could even place it in the `common` library given that it will be used by multiple classes.

jreback replied:
you could modify ``_validate_kwargs`` (maybe call it ``_validate_arguments`` or something), and do both I think

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +1009,94 @@ def factorize(self, sort=False, na_sentinel=-1):
+        -----
+        Binary search is used to find the required insertion points.
+
+        The 'sorter' argument is not used for 'Index' instances, so a
```
not a helpful comment

gfyoung replied:
Are you referring to the `sorter` argument one?  If so, I don't quite understand why that isn't helpful, since it would alert the user about the `NotImplementedError`.

jreback replied:
if we are actually going to leave the arg in, then a note that a particular arg will raise ``NotImplementedError`` if its not ``None`` should certainly be WITH the description of the argument not later

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +1009,94 @@ def factorize(self, sort=False, na_sentinel=-1):
+        array([1])
+        >>> x.searchsorted(['bread', 'eggs'])
+        array([1, 4])
+        >>> x.searchsorted(['bread', 'eggs'], side='right')
```
nice examples!

gfyoung replied:
Not going to lie, but I ported those over from the doc strings for `Series.searchsorted` and `Categorical.searchsorted`.  Note that some of the function calls pass in a `sorter` argument.

jreback replied:
no I get that. but still need to review & edit.

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +1009,94 @@ def factorize(self, sort=False, na_sentinel=-1):
+        array([3, 4])    # eggs before milk
+        >>> x = pd.Categorical(['apple', 'bread', 'bread', 'cheese', 'milk',
+                                'donuts' ])
+        >>> x.searchsorted(['bread', 'eggs'], side='right',
```
we are not allowing ``.sorter``

gfyoung replied:
See my response above regarding `Series.searchsorted`.

jreback replied:
read my response above.

jreback commented on pandas/tests/test_base.py:
```diff
@@ -972,6 +972,23 @@ def test_memory_usage(self):
+            index = np.searchsorted(o, max(o))
+            self.assertTrue(0 <= index <= len(o))
+
+            if isinstance(o, Index):
```
assert for all types

gfyoung replied:
I presume you must be referring to the fact that `sorter` should be `None` in all cases.  See my comment above regarding `Series.searchsorted`.

gfyoung commented:
I am testing out hiding the 'sorter' argument.  This is WIP.  There are some instances in the code base where the `sorter` argument is actually used (e.g. `tests/test_categorical.py` and `core/series.py`).  Should I just remove them?  And if so, is this more of an API change than a bug fix?

gfyoung commented:
I am testing out hiding the 'sorter' argument.  This is WIP.  There are some instances in the code base where the `sorter` argument is actually used (e.g. `tests/test_categorical.py` and `core/series.py`).  Should I just remove them?  And if so, is this more of an API change than a bug fix?

shoyer commented on pandas/core/base.py:
```diff
@@ -990,13 +991,73 @@ def factorize(self, sort=False, na_sentinel=-1):
+
+    @Substitution(klass='IndexOpsMixin')
+    @Appender(_shared_docs['searchsorted'])
+    def searchsorted(self, v, side='left', *args):
```
Let's call this `*unused_args` rather than `*args`. And let's also add `**unused_kwargs` in case something like your NumPy PR ever gets merged switching this to a keyword argument.

gfyoung replied:
Sure thing.  If there are no complaints for that PR in `numpy`, could you merge that in at some point?

shoyer commented on pandas/core/base.py:
```diff
@@ -990,13 +991,73 @@ def factorize(self, sort=False, na_sentinel=-1):
+        # needs coercion on the key (DatetimeIndex does already)
         # needs tests/doc-string
-        return self.values.searchsorted(key, side=side)
+        com.validate_args(args, min_length=0, max_length=1)
```
This still needs to raise an exception if `args != (None,)` (the default value for `sorter`)

gfyoung replied:
Why?  The idea was that we swallow the `sorter` arg without saying anything.  No need for an `Exception`.  That's how it has been done for other `numpy`-compat functions where the `**kwargs` swallowed arguments.

shoyer replied:
You're probably right that there are cases where we currently use `**kwargs` to swallow arguments, but this is something we should improve in new code :). e.g., for `.mean()`, we should probably be writing something like this:
```
def mean(self, axis, skipna=True, **unused_kwargs):
    if unused_kwargs.get('keepdims', False):  # False is the default value from NumPy
        raise ValueError(...)
```

From a user experience point of view, I think it's worse to swallow invalid arguments than it is to always raise unexpected errors. In the later case, users will quickly realize that `np.searchsorted` doesn't work on pandas objects. But if we silently accept invalid values, they'll think `np.searchsorted` works until much later they may discover their code was giving erroneous results. We try to avoid these types of bugs :).

gfyoung replied:
Fair enough, though I think the change would have to be more gradual, since if the issue I'm fixing is any indication, many people I suspect might believe `np.searchsorted` (and various other `numpy` functions) should "work" with `pandas` objects.  In fact, that is the idea behind several `pandas` functions that have counterparts in `numpy` as far as I have seen.

This may be another issue you might want to discuss with @jreback , since he has generally advocated a more `numpy` friendly interface for `pandas` so that more `numpy` dependent users can use `pandas` more easily given how similar the interface may appear at the surface.

At this point, IMO it would be best to save the `Exception` raising for a separate PR in which the interface would change across the board to become more `numpy`-hostile (for lack of a better term), as this does feel like a somewhat significant change.

shoyer replied:
Indeed, I'm totally +1 on pandas being as numpy friendly as possible.

*However*, it's **not** numpy friendly to accept but ignore additional arguments if they have non-default values. Are you seriously saying that `np.searchsorted(pandas_obj, sorter=other)` should work by ignoring `other`?

gfyoung replied:
Again, I think it would be best to discuss this with @jreback , as this is a question of overall API design for the repository.

shoyer commented:
We don't want to remove existing cases where `sorter` works. If anything, I would make `sorter` work in more places (instead of requiring that the argument always be `None`).

gfyoung commented:
I'll leave you to discuss that point with @jreback .  He seems to be of the opposite camp that the `sorter` argument doesn't add much for the `pandas` implementation.

jreback commented on pandas/core/common.py:
```diff
@@ -3095,3 +3095,32 @@ def pandas_dtype(dtype):
+
+
+def validate_args(args, min_length=0, max_length=None):
+    """
```
let's move ``core/generic.py/_validate_kwargs`` here as well (and put a comment in this seconds of 2 functions to indcate the purpose)

gfyoung replied:
Could you clarify your statement?  I didn't quite understand the parenthetical.

jreback replied:
typo! I meant put a comment in ``core/common.py`` to indicate that this is a sub-section (for those 2 functions). ``core/common.py`` needs reorg, see #12503, and this will be an indication that these 2 functions belong together (you can also take suggestion from that issue and go ahead and put them in a new sub-module)

gfyoung commented:
@jreback : Did the refactoring, and tests are passing.

jreback commented on pandas/util/validators.py:
```diff
@@ -0,0 +1,48 @@
+"""
+
+
+def _validate_args(args, min_length=0, max_length=None):
```
rename these w/o the leading _ eg just validate_args and same for validate_kwargs

gfyoung replied:
Sure thing.  Done.

jreback commented:
minor naming change - otherwise lgtm

shoyer commented:
As I mentioned inline above, I think this needs to raise an exception for an invalid `sorter` argument. Existing precedent for buggy APIs is not a good excuse for making new buggy designed APIs. @jreback, your thoughts?

gfyoung commented:
@jreback : Renamed the functions, and Travis remains happy.

jreback commented:
will give another look later, but my general philosophy is in-line with @shoyer maybe it didn't come across.

We don't want to advertise ``sorter`` when used with ``s.searchsorted(...)`` IOW a pandas method. But don't want to disallow passing with ``np.searchsorted(.....)``. So I suppose we can use completely ignore the ``sorted`` argument (if passed), as its not clear when its passed for compat reasons (e.g. from INSIDE ``np.searchsorted``), or the user passing it in. but not sure it really matters.

gfyoung commented:
@jreback : Sure thing.  I think the part that seems conflicting between you and @shoyer is the handling of `numpy` arguments when they are not their default values.  While you seemed to have preferred to just ignore it (that was the impression I got when I wrote the `_validate_kwargs` method in a previous PR), @shoyer is advocating throwing an `Exception` instead.

While I don't know how `searchsorted` will work in the future, but FYI, the way it works when you call `np.searchsorted` is that it checks whether the array-like object being passed in has its own `searchsorted` method.  If it does, it calls that method using the signature `(val, side=`left`, sorter=None)`.  If it does not have such a method, the object is wrapped in a `numpy` array and that array's `searchsorted` method is called instead.

jreback commented on pandas/core/series.py:
```diff
@@ -1462,63 +1463,21 @@ def dot(self, other):
-        Find the indices into a sorted Series `self` such that, if the
-        corresponding elements in `v` were inserted before the indices, the
-        order of `self` would be preserved.
+        if not isinstance(self._values, Index):
```
why do we need this again? (e.g. the explicit check if its an ``Index``)

gfyoung replied:
That's the same question I asked myself when I looked at it.  Not sure how / why it was ever put there.  I'll remove the check.

jreback commented on pandas/tests/series/test_analytics.py:
```diff
@@ -1625,7 +1625,7 @@ def test_search_sorted_datetime64_list(self):
         # GH8490
         s = Series([3, 1, 2])
-        r = s.searchsorted([0, 3], sorter=np.argsort(s))
+        r = s.searchsorted([0, 3], 'left', np.argsort(s))
```
add a comment that this is for compat (and reference the numpy issue, and this one), about why we are passing args and not kwargs

gfyoung replied:
Doesn't that come into conflict with your suggestion about adding `**kwargs` to `searchsorted`?

jreback commented on pandas/tests/test_base.py:
```diff
@@ -972,6 +972,18 @@ def test_memory_usage(self):
+
+            # check that the 'sorter' argument can be
+            # passed in via numpy's searchsorted call
+            index = np.searchsorted(o, max(o), sorter=range(len(o)))
```
any exceptions to tests here?

gfyoung replied:
Not sure if I do in the context of compatibility enforcement.  This test first tests that behaviour is well-defined when you "correctly" call `np.searchsorted` (i.e. with only the arguments `pandas` wants) and then "incorrectly" (i.e. with all the arguments that `numpy` would accept), so I think I've covered all the necessary `numpy` compat cases IMO.

jreback replied:
just add a test with a sorter arg that is invalid

gfyoung replied:
Fair enough.  Will do.

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -1410,10 +1410,8 @@ def test_searchsorted(self):
-                              sorter=[0, 1, 2, 3, 5, 4])
-        chk = s2.searchsorted(['bread', 'eggs'], side='right',
-                              sorter=[0, 1, 2, 3, 5, 4])
+        res = c2.searchsorted(['bread', 'eggs'], 'right', [0, 1, 2, 3, 5, 4])
```
ok for ``side`` to be a kwarg right?

gfyoung replied:
You mean pass it in in this line as a keyword instead of a positional argument?

jreback replied:
yes

gfyoung replied:
Oh, you know why: can't have `**kwargs` before `*args` - that's why I had to change it!

jreback replied:
you simply need both

gfyoung replied:
What do you mean by "both"?

jreback replied:
the signature needs both

gfyoung replied:
Oh you mean this:
~~~{.Python}
def searchsorted(..., side='left', *args, **kwargs):
    ...
~~~
instead of just
~~~{.Python}
def searchsorted(..., side='left', *args):
    ...
~~~

jreback replied:
yep

gfyoung replied:
I feel somewhat concerned about this.  Doesn't this make things more complicated?  If we're trying to discourage usage of the `sorter` argument, it seems little strange that we're giving the user so many ways to pass it in.

jreback replied:
not at all. you understand I don't want the ``sorter`` arg AT ALL. But we are being forced to have it passed in because of numpy. So when numpy starts sending it with ``**kwargs`` instead of as a named argument just accepting ``*args`` isn't enough. Might as well just remove it now. Its really not big deal to do.

gfyoung replied:
Well, I took your suggestion about the `try`-`except` to heart in my `numpy` PR so `numpy` won't ever be sending it in as `**kwargs` unless someone beats me to it.

jreback commented on pandas/util/validators.py:
```diff
@@ -0,0 +1,48 @@
+    has at least `min_length` arguments. If `max_length` is an integer, checks
+    whether `*args` has at most `max_length` arguments inclusive. Raises a
+    ValueError if any of the conditions are False.
+    """
```
can you add a test module to (in ``pandas.util``) to validate these validators, just some simple tests that exercise each path.

gfyoung replied:
Sure thing.

jreback commented:
I know we have beat this to death! looks pretty good. Just a comment about why special casing the check for ``sorter``, further why are we not accepting ``**kwargs`` (in addition to ``*arg``) for ``.searchsorted()``? (e.g. make sure it is empty).

shoyer commented on pandas/tseries/index.py:
```diff
@@ -1629,13 +1632,17 @@ def normalize(self):
+    @Substitution(klass='DatetimeIndex')
+    @Appender(_shared_docs['searchsorted'])
+    def searchsorted(self, v, side='left', *args):
+        validate_args(args, min_length=0, max_length=1)
```
we're still intentionally ignoring `sorter` here?

gfyoung replied:
AFAICT, yes?  @jreback ?

gfyoung replied:
@shoyer : I have been working on both ends of this compatibility issue, and from what I can see, it does seem unfortunate that `numpy` "requires" the implementation of methods like `searchsorted` in downstream libraries to match that of `numpy`'s.  My PR against `numpy` is going to make it lenient with this (see <a href="https://github.com/numpy/numpy/pull/7325">#7325</a>) by wrapping the object in an array whenever an `AttributeError` (i.e. object has no such method) or a `TypeError` (i.e. object has such a method but a different signature) is thrown.

This would help to relieve any of these weird compatibility issues on the `pandas` end because we could then develop these methods <u>independently</u> of `numpy`'s without having to pass in these extraneous `*args` and `**kwargs` params in for them.

jreback replied:
``pandas`` will not create one, but if you specify it directly using ``np.searchsorted`` then we are passing it thru. At least that's the intention (and looks like code/tests exercise that).  So we are passing thru numpy args that we don't care about, but won't error on them (as its very hard to tell *where* they actually come from), e.g. a user could in theory

``s.searchsorted(....., sorter=[....])``. But since its not in the doc-string, nor signature its not likely.

The exception would be in cases where we are not allowing something (e.g. in the ``np.round`` case, where ``out=`` is passed in intending to be used to do something in-place, which we don't allow via that API). So that *should* raise (if ``out`` is not None, but that's the other issue).

gfyoung replied:
@jreback : Regarding args that are accepted in the `numpy` implementation but are extraneous to the `pandas` implementation`, where do you draw the boundary between throwing when they are non-null and not (i.e. we throw when `out != None` but not when `sorter != None` as was proposed by @shoyer)?

jreback replied:
easy, ``out`` is the user expecting an in-place modification, which we don't do that way, so we MUST raise (prob ``ValueError``). while ``sorter`` we don't care whether its passed or not and so it gets passed thru to satisfy numpy compat.

gfyoung replied:
But if you pass in a non-null argument for `sorter`, the user is expecting the array-object to be arranged in that order FIRST and then be searched.  While I do agree that `out` is more of a serious expectation, the cutoff feels a little fuzzy to me.

gfyoung commented on pandas/core/series.py:
```diff
@@ -1462,63 +1463,21 @@ def dot(self, other):
+        if not isinstance(self._values, Index):
+            if sorter is not None:
+                sorter = com._ensure_platform_int(sorter)
 
```
@jreback, @shoyer: Is there any reason why `Series` was the only class that really supported `sorter`?  I know that its `_values` attribute is a `numpy` array, but in the spirit of the other classes, does it make sense just to remove this?  It seems somewhat inconsistent to try to maintain this but drop support of `sorter` everywhere else.

jreback replied:
we DON"T support ``sorter`` AT ALL!. It was just there as copied originally from numpy. Don't make this more compicated. Just remove the argument and let's be done with this. This should be really simple, and is dragging on.

gfyoung replied:
Okay, okay.  Got it.  Will remove it as well.

jorisvandenbossche replied:
sorry for the back and forth, but, @jreback you may like it or not, but we *do* support `sorter` for `pd.Series.searchsorted`: it is documented in the docstring and it is explicitly passed through to numpy's searchsorted

So we either remove it, but then it's an API change and should wait until 0.19, or we just keep it. And to be honest, I don't see the problem in keeping it. It is just simply passing it to numpy's method (it does not make the code any more complicated), and we have to accept it anyway regardless of if we pass it through or not (for the compat issues that started this PR)

jreback replied:
it's a non supported argument that serves no purpose and removing it is not an API change that is visible - it will still be accepted just not advertised so I see no reason not to clean this up now

shoyer replied:
I agree with Joris. It's documented in the API docs:
http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.searchsorted.html

On Mon, Mar 14, 2016 at 6:14 PM, Jeff Reback <notifications@github.com>
wrote:

> In pandas/core/series.py
> <https://github.com/pydata/pandas/pull/12413#discussion_r56102581>:
>
> >
> > -        Find the indices into a sorted Series `self` such that, if the
> > -        corresponding elements in `v` were inserted before the indices, the
> > -        order of `self` would be preserved.
> > +        if not isinstance(self._values, Index):
> > +            if sorter is not None:
> > +                sorter = com._ensure_platform_int(sorter)
> >
>
> it's a non supported argument that serves no purpose and removing it is
> not an API change that is visible - it will still be accepted just not
> advertised so I see no reason not to clean this up now
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/12413/files#r56102581
>

jreback replied:
@shoyer you have flip flopped

shoyer replied:
It's documented in the API docs for Series, not Index. We have three options here I would be OK with here:

- add `sorter` to Index as a valid argument (slight preference)
- deprecate `sorter` from Series (but save the API change for later)
- don't change the status of `sorter` on Index, leaving things slightly inconsistent for now

No matter what, it would be a bad idea to start silently accepting but ignoring the `sorter` arguments. That is the way of bugs. If we're going to add `sorter` but not accept it as a real argument, we need to verify that `sorter is None` or raise an error.

jorisvandenbossche replied:
> That is the way of bugs. If we're going to add sorter but not accept it as a real argument, we need to verify that sorter is None or raise an error.

I agree that we should raise an error in case of ignoring the keyword. 
But for that reason, it's actually much simpler of just passing it through to numpy, then we don't have to care about it further.

I don't see a compelling reason to start deprecating it, so my preference is the first bullet point as well, or the third.

So

jreback commented on pandas/tests/test_util.py:
```diff
@@ -73,6 +75,66 @@ def test_rands_array():
+
+class TestValidateArgs(tm.TestCase):
+
+    def test_bad_min_length(self):
```
gr8!

jreback commented on pandas/util/validators.py:
```diff
@@ -0,0 +1,48 @@
+def validate_args(args, min_length=0, max_length=None):
+    """
+    Checks whether the length of the `*args` argument passed into a function
+    has at least `min_length` arguments. If `max_length` is an integer, checks
```
can you make these have the ``Parameters`` section (like we doc arguments) & Returns.

gfyoung replied:
Not sure that there actually is a `Returns` section in the documentation as far as I understand but can definitely add a `Parameters` section.

jreback replied:
speaking in general, you can list ``Raises`` which is more useful here

gfyoung replied:
Ah, good point.  Will do.

jreback commented on pandas/util/validators.py:
```diff
@@ -0,0 +1,48 @@
+
+def validate_kwargs(fname, kwargs, *compat_args):
+    """
+    Checks whether parameters passed to the **kwargs
```
same here

gfyoung replied:
:+1:

gfyoung commented:
Updated PR with your comments regarding the documentation.  Added `[ci skip]` since it is code-wise no different from build <a href="https://travis-ci.org/pydata/pandas/builds/115989327">#18286</a>, which I am still waiting to pass, though it should most likely pass since it passed locally.

jreback commented:
ok, so to lay out a policy here which is compat with several goals.

- we don't really need or want the ``sorter`` argument, but ``numpy`` requires it (as a positional). So am forced to accept it and document it as an argument that we just pass thru. So ``Series`` is good now, @gfyoung let's keep the modification where you put it (IOW have ``Categorical/Index``) support this so its all in 1 place. Remove the ``*args, **kwargs`` checking. You can put back minimal tests for passing this thru (numpy will do all error checking). Make a note in the doc-string that this is just a pass thru argument.

The reasoning here is that we have a positional argument that we are forced to accept for compat reasons. We cannot easily ignore this w/o jumping thru a bunch of hoops. We are forced to have this in the doc-string.


- other PR (about ``np.round``) and stat functions. We *need* to accept ``out``, but explicity don't do anything or want this argument. Here let's check if its ``None`` (its in ``**kwargs`` already), if not raise a ``TypeError`` (we do this already). Further goal is since we are accepting ``**kwargs`` need to make sure that we dont take addtl args that are not in the doc-string.

The reasoning here is we have a kwargs argument that we don't want (and don't want in the doc-string), and can safely make sure its ``None``


Bottom line is that we don't want to have arbitrary / odd argument in the doc-strings (and signatures) just for compat, unless we have to jump thru hoops to get rid of them. Having arguments which don't do anything is just bad practice in general and makes things even more confusing. 

In particular I am (I guess was) making a case that ``sorter`` is a really numpy specific arg that in pandas space is completely useless and misleading (we NEVER need this, we ALREADY have a concept of an index and values; numpy NEEDS this because it doesn't).

I think it is worthwhile to hide completely numpy specific useless things from the user. We do this for many many other cases. So philosophically I don't see any reason to digress here. The reason I am yielding is that there *is* a slight usecase to support ``np.searchsorted(...)`` and the fact that numpy passes these positionally BACK to use (with ``_wrap_it``), is just plain confusing (and @gfyoung has already opened a report on this, but this can't affect our decision here because of back-compat with older numpies).

We have in the past few years been more and more hiding numpy from end-user for normal cases. I think this is just going to increase to where you won't be able to TELL if numpy is there or not, it just won't matter at all. Therefor anything that can reduce friction with the end-user and reduce confusion is a good thing, even at the expense of some work-arounds.

gfyoung commented:
Alright, let me see if I understand this correctly:

For compatibility reasons, we are going to add `sorter` as a **keyword** argument for all relevant data structures (i.e. `Index`, `Category`, `DataFrame`, etc...).  However, for the majority of these data structures (excluding `Series`), this will serve as a **pass thru** argument that will be validated as per @shoyer.  For `Series`, I am personally in favor of deprecating `sorter` because it's going to be confusing for the user if we support it for one data structure but not for the others.

In this case, I'll remove the `*args` and `**kwargs` checking and move that into my `np.round` PR and remove its dependency on this one.  Also, @jreback , unfortunately, it will have to be an `*args` thing because `numpy` passes it in internally as a positional argument despite accepting as a keyword!

jreback commented:
@gfyoung maybe I wasn't clear.

- ``.searchsorted()``, centralize in 1 place for everything (you already did this). add back ``sorter`` arg, remove ``*arg, **kwargs``. no validation, it just gets passed thru (it will be a named argument, defaulting to ``None``, this will work with how numpy is passing it now).
- ``.round()`` (and stats): validation as currently exists (in the stats functions for ``out``, add for ``.round()``) (but use the new validation routines you cooked up).

jorisvandenbossche commented:
> - ``.searchsorted()``, centralize in 1 place for everything (you already did this). add back ``sorter`` arg, remove ``*arg, **kwargs``. no validation, it just gets passed thru (it will be a named argument, defaulting to ``None``, this will work with how numpy is passing it now).
- ``.round()`` (and stats): validation as currently exists (in the stats functions for ``out``, add for ``.round()``) (but use the new validation routines you cooked up).

+1

jorisvandenbossche commented:
@gfyoung DataFrame has no searchsorted, so it is only series and index (and categorical I suppose)

gfyoung commented:
Another reason for supporting deprecation of `sorter` in `Series`: unpredictable behaviour.  I just got test failures because the internal data-structure (i.e. the `_values` attribute) was an `Index` and not a `numpy` as I had initially assumed to be the case.

jreback commented:
it's not unpredictable but based on the dtype 
your assumption is not correct

gfyoung commented:
Fair enough, but what that fact means is that whether or not `Series` actually respects `sorter` is dependent on the `dtype` of its internal `_values` attribute, which makes things more complicated for the user and adding further reason to deprecate it.

gfyoung commented:
That point aside, a new draft of the PR is up now.  Comments are welcome!

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -60,7 +60,8 @@ API changes
 
-
+- The ``sorter`` argument for ``searchsorted`` has been deprecated for ``Series`` and ``Categorical`` objects,
+  as its existence is largely for numpy compatibility purposes. It will be removed in the future. (:issue:`12238`)
```
no, not deprecating.

jreback replied:
remove this

gfyoung replied:
Done.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -89,7 +90,6 @@ Bug Fixes
 - Bug in ``value_counts`` when ``normalize=True`` and ``dropna=True`` where nulls still contributed to the normalized count (:issue:`12558`)
-
 - Bug in ``CategoricalIndex.get_loc`` returns different result from regular ``Index`` (:issue:`12531`)
+- ``searchsorted`` now accepts a ``sorter`` argument to maintain compatibility with numpy's ``searchsorted`` function (:issue:`12238`)
```
you need to be specific here. its ``Index/Categorical`` that is now acccepting it to be compat with ``Series``

gfyoung replied:
But I did add `sorter` as an argument for all of the other implementations of `searchsorted` in the codebase?

jreback replied:
and these are ``Index`` and ``Categorical``. A user would have no idea what you are tallking about.

gfyoung replied:
Done.

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +991,87 @@ def factorize(self, sort=False, na_sentinel=-1):
+                              "storing ndarray, Categorial, or Series data."))
+
+        # needs coercion on the key (DatetimeIndex does already)
+        return self.values.searchsorted(v, side=side)
```
might as well pass sorter in

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +991,87 @@ def factorize(self, sort=False, na_sentinel=-1):
+            order. They are typically the result of ``np.argsort``. If `self`
+            is not a ``Series`` object, the default value of `sorter` should
+            always be passed in because it is only there for compatibility
+            purposes with NumPy. If the default value is not passed in, a
```
make this much shorter.

gfyoung replied:
Most of that doc-string was due to the deprecation and all that jazz.  Removed.

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +991,87 @@ def factorize(self, sort=False, na_sentinel=-1):
+            However, such usage is **now deprecated**, and support for the
+            argument will be removed entirely in the future.
+
+            .. deprecated:: 0.18.1
```
we don't use these tags

gfyoung replied:
Ah, good to know.  Removed since there no longer is any deprecation.

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +991,87 @@ def factorize(self, sort=False, na_sentinel=-1):
+    @Appender(_shared_docs['searchsorted'])
+    def searchsorted(self, v, side='left', sorter=None):
+        if sorter is not None:
+            raise ValueError(("Non-default values for 'sorter' are only "
```
take this out

jorisvandenbossche commented:
> Another reason for supporting deprecation of sorter in Series

Or you can also put that as: that supports the case for including it in Index.searchsorted

I am personally still in favor of @shoyer first option. Repeating the options here as it is a bit hidden in one of the code comments:

1. add `sorter` to Index as a valid argument (slight preference)
2. deprecate `sorter` from Series (but save the API change for later)
3. don't change the status of `sorter` on Index, leaving things slightly inconsistent for now

jreback commented on pandas/core/categorical.py:
```diff
@@ -1003,59 +1005,13 @@ def memory_usage(self, deep=False):
+    @Appender(_shared_docs['searchsorted'])
     def searchsorted(self, v, side='left', sorter=None):
-        """Find indices where elements should be inserted to maintain order.
+        if sorter is not None:
```
take this out

jreback commented on pandas/core/series.py:
```diff
@@ -1462,63 +1462,16 @@ def dot(self, other):
         if sorter is not None:
-            sorter = com._ensure_platform_int(sorter)
+            warnings.warn(("The 'sorter' argument is now deprecated and "
+                           "will be removed in a future version."),
```
take this out

jreback commented on pandas/tests/test_base.py:
```diff
@@ -972,6 +972,34 @@ def test_memory_usage(self):
+            index = np.searchsorted(o, max(o))
+            self.assertTrue(0 <= index <= len(o))
+
+            msg = ("Non-default values for \'sorter\' are only "
```
fix this

jreback commented on pandas/tests/test_base.py:
```diff
@@ -972,6 +972,34 @@ def test_memory_usage(self):
+            index = np.searchsorted(o, max(o))
+            self.assertTrue(0 <= index <= len(o))
+
+            with tm.assert_produces_warning(DeprecationWarning):
```
remove this (and FYI we don't ever use ``DeprecationWarning``). We use ``FutureWarning``

jreback commented on pandas/tseries/index.py:
```diff
@@ -1629,13 +1631,20 @@ def normalize(self):
+    def searchsorted(self, v, side='left', sorter=None):
+        if sorter is not None:
+            raise ValueError(("Non-default values for 'sorter' are only "
+                              "accepted for Categorical and Series objects "
```
fix this

jorisvandenbossche commented:
And as I understand it correctly, that was also what @jreback summarized above:

> - ``.searchsorted()``, centralize in 1 place for everything (you already did this). add back ``sorter`` arg, remove ``*arg, **kwargs``. no validation, it just gets passed thru (it will be a named argument, defaulting to ``None``, this will work with how numpy is passing it now).

Add back sorter, just pass through, no deprecation.

jreback commented:
pls remove all of the deprecation stuff. Just make this simple and we'll get it in.

jorisvandenbossche commented on pandas/core/base.py:
```diff
@@ -990,13 +991,87 @@ def factorize(self, sort=False, na_sentinel=-1):
+        sorter : 1-D array_like, optional
+            Optional array of integer indices that sort `self` into ascending
+            order. They are typically the result of ``np.argsort``. If `self`
+            is not a ``Series`` object, the default value of `sorter` should
```
This is way too complicated for in a user docstring. There is no notion of `self` for the user in the docstring.

Another reason for just accepting it for all types, and just pointing to the numpy docs for its meaning

jreback replied:
I had same comment above, lol

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -114,6 +114,7 @@ Bug Fixes
 
 
 - Bug in ``CategoricalIndex.get_loc`` returns different result from regular ``Index`` (:issue:`12531`)
+- ``searchsorted`` for ``Index`` and ``Categorical`` now accept a ``sorter`` argument to maintain compatibility with numpy's ``searchsorted`` function (:issue:`12238`)
```
let's put in API changes section

gfyoung replied:
Done.

jreback commented on pandas/core/categorical.py:
```diff
@@ -1064,7 +1016,9 @@ def searchsorted(self, v, side='left', sorter=None):
-        return self.codes.searchsorted(values_as_codes, sorter=sorter)
+
+        return self.codes.searchsorted(values_as_codes,
+                                       side='left', sorter=sorter)
```
``side=side``

gfyoung replied:
Ah, good point.  Done.

jreback commented:
small correction (and pls add a test to catch that in the categorical tests).

otherwise lgtm

@jorisvandenbossche @shoyer

gfyoung commented:
@jreback : Since we're accepting and using `sorter` in `base.py` when we weren't before, any reason why we shouldn't be adding it to `period.py` and `tdi.py`.

Also, FYI, `Categorical` actually was already accepting `sorter`.  It was `TimedeltaIndex` that was not before, so I updated the API comment to reflect that.

jreback commented:
yes they should all pass thru

gfyoung commented:
@jreback : Also, regarding that `side` typo I made with `Categorical` initially, I don't think that was ever an issue because the line before it made sure that the `side` parameter was enforced.

jreback commented:
@gfyoung ok then

jreback commented on pandas/core/base.py:
```diff
@@ -990,13 +991,73 @@ def factorize(self, sort=False, na_sentinel=-1):
+
+        Parameters
+        ----------
+        v : array_like
```
I know this is annoying, but how about we rename ``v`` -> ``value``. More readable args.

gfyoung replied:
In a previous comment we changed it to `v` so that it matches the signature of `numpy`, or maybe I misunderstood you then?

jreback replied:
no I think it was always like this. But we generally name things as ``value`` in pandas (single character is too like q/kdb .... )

jorisvandenbossche replied:
Let's maybe just keep it at `v` (as long as we do it better in our own methods!), you never know that someone used it as a keyword argument

gfyoung replied:
@jorisvandenbossche : `v` is not used as a keyword in `numpy`, and from a semantic point of view, I see no reason why it would be.

jorisvandenbossche replied:
But you *can* use it as a keyword, like `pd.Series([1,2,3]).searchsorted(v=1)`, which would then break. So IMO this is an unneeded API change

jreback replied:
ok its fine. just wanted consistency with other pandas functions. let's make an issue and change it for 0.19.0.

gfyoung replied:
@jorisvandenbossche : I think you may have hit a Python bug in fact.
~~~.{Python}
>>> class Demo(object):
             def illustrate(self, a, b=2, c=3):
                   print("See, I worked, and 'a' is not a keyword argument.")
>>> demo = Demo()
>>> demo.illustrate(a=5)
"See, I worked, and 'a' is not a keyword argument."
~~~
This is a Python bug, which unfortunately we will have to live with, so yes, `values` should become `v` again...

kawochen replied:
@gfyoung it looks correct to me.  what bug does this demonstrate (and you are indeed calling `illustrate` with a keyword argument)?

gfyoung replied:
@kawochen : Maybe I should have known this before, but you can pass in positional arguments as keyword arguments?

kawochen replied:
Keyword arguments on the calling side just mean that you can specify which named argument your argument goes into as.  This is useful when you want to list out you arguments in a different order than in the function definition.

gfyoung replied:
Yikes, this is news to me.  I guess I never encountered that before.  Well, at least I know now and am slightly less of a Python noob because of that.  Thanks for clarifying @kawochen!

jreback replied:
ok, just about to merge, then I saw you changed *some* of the ``v`` -> ``value``. (you missed the datetimelikes). 

So originally we had datetimelikes used ``key`` (and I had you change it to ``v``). While ``Series`` used ``v``.

So we have to break something.

Let's just keep the rename to ``value`` (and fix the datetimelikes). Move the whatsnew note to ``API Changes`` (and note that the argument name changed). I seriously doubt *anyone* is actually passing as a kw arg (the first arg).

gfyoung replied:
Okay, so rename to `value` and make a note in the API changes?  Also, what do you mean by "datetimelikes"?

jreback replied:
datetimes, timedelta, period

gfyoung replied:
Got it.  Done.

gfyoung commented:
@jreback : Changes have been made, and Travis approves.  Should be good to merge then.

jorisvandenbossche commented:
Sorry, but I want to say it again. Changing a keyword's name is an API change that can break people's code, and since we want to change it because "we don't like the short name", it is certainly not an unavoidable API change, so IMO should not go into a minor release.

gfyoung commented:
@jorisvandenbossche : I find myself caught in a tug-of-war between you and @jreback on this issue, not to mention the push to land this is very high.  It would be great if some consensus could be found on this.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -43,8 +43,8 @@ Enhancements
-
-
+- ``searchsorted`` for ``Index`` and ``TimedeltaIndex`` now accept a ``sorter`` argument to maintain compatibility with numpy's ``searchsorted`` function (:issue:`12238`)
+- The parameter for specifying values to be inserted into a ``pandas`` object when calling its ``searchsorted`` method is now uniformly denoted as ``value``
```
pls just make the whatsnew simpler. just say the signature changed the names of the arguments and put in a pointer to the signature.

gfyoung replied:
Done.  I put `[ci skip]` in my commit because this is only a doc change.  So it should be good to merge if there is nothing else.

jreback commented:
@jorisvandenbossche well, we already have that issue as ``key`` is the name for the datetimelikes and ``v`` for Series. I guess its ok to move this part of the change to 0.19.0 (the actual name changes). Though really don't think this is a big deal to change this now as its the first argument.

jorisvandenbossche commented:
But I think it is also not a big deal to just leave it for 0.19.0, just in case to be safe. But I agree the discrepancy with Index `key` is indeed also not nice.

@gfyoung would you like to split the PR in two? one for the actual fix to searchsorted and one for the name change?

jreback commented:
ok, let's move the name change to 0.19.0. (you can just create an issue I think and I'll mark it). otherwise ok then.

gfyoung commented:
@jreback : Travis has finally given the green light on this one as well.  Should be good to merge if there is nothing else to add.

jreback commented:
whoosh! thanks for the patience

gfyoung commented:
@jreback: No worries.  Onto fixing the rest of `fromnumeric.py`!
